tools/kagami/src/swarm.rs

```diff
@@ -0,0 +1,1314 @@
+                    peer,
+                    service_source.clone(),
+                    volumes.clone(),
+                    trusted_peers.clone(),
+                    None,
+                );
+
+                (name.clone(), service)
+            })
+            .chain(std::iter::once(first_peer_service))
+            .collect();
+
+        let compose = DockerCompose::new(services);
+        Ok(compose)
+    }
+}
+
+#[derive(Clone, Debug)]
+struct AbsolutePath {
+    path: PathBuf,
+}
+
+impl Deref for AbsolutePath {
+    type Target = PathBuf;
+
+    fn deref(&self) -> &Self::Target {
+        &self.path
+    }
+}
+
+impl AsRef<Path> for AbsolutePath {
+    fn as_ref(&self) -> &Path {
+        self.path.as_path()
+    }
+}
+
+impl AbsolutePath {
+    fn absolutize(path: PathBuf) -> Result<Self> {
+        Ok(Self {
+            path: if path.is_absolute() {
+                path
+            } else {
+                path.absolutize()?.to_path_buf()
+            },
+        })
+    }
+
+    /// Relative path from self to other.
+    fn relative_to(&self, other: &AbsolutePath) -> Result<PathBuf> {
+        pathdiff::diff_paths(self, other)
+                .ok_or_else(|| {
+                    eyre!(
+                        "failed to build relative path from {} to {}",
+                        other.display(),
+                        self.display(),
+                    )
+                })
+                // docker-compose might not like "test" path, but "./test" instead 
+                .map(|rel| {
+                    if rel.starts_with("..") {
+                        rel
+                    } else {
+                        Path::new("./").join(rel)
+
+                    }
+                })
+    }
+}
+
+mod key_gen {
+    use iroha_crypto::{error::Error, KeyGenConfiguration, KeyPair};
+
+    /// If there is no base seed, the additional one will be ignored
+    pub fn generate(base_seed: Option<&[u8]>, additional_seed: &[u8]) -> Result<KeyPair, Error> {
+        let cfg = base_seed
+            .map(|base| {
+                let seed: Vec<_> = base.iter().chain(additional_seed).copied().collect();
+                KeyGenConfiguration::default().use_seed(seed)
+            })
+            .unwrap_or_default();
+
+        KeyPair::generate_with_configuration(cfg)
+    }
+}
+
+mod peer_generator {
+    use std::{collections::BTreeMap, num::NonZeroU16};
+
+    use color_eyre::{eyre::Context, Report};
+    use iroha_crypto::KeyPair;
+    use iroha_data_model::prelude::PeerId;
+    use iroha_primitives::addr::{SocketAddr, SocketAddrHost};
+
+    const BASE_PORT_P2P: u16 = 1337;
+    const BASE_PORT_API: u16 = 8080;
+    const BASE_PORT_TELEMETRY: u16 = 8180;
+    const BASE_SERVICE_NAME: &'_ str = "iroha";
+
+    pub struct Peer {
+        pub name: String,
+        pub port_p2p: u16,
+        pub port_api: u16,
+        pub port_telemetry: u16,
+        pub key_pair: KeyPair,
+    }
```

Comment:
Seeing that it's pretty much the `test_network` peer struct without the optional fields, one part of me wants to remove the redundancy, while I also don't see a way to do it..

Reply:
This struct has only several fields and functionality, and its purpose here is very limited: peer representation in the docker compose file. So, I think that introducing an additional modules coupling (`kagami::swarm` ⭤ `test_network`) here will bring more headache than improvement.